arangod/Aql/Executor/EnumerateNearVectorExecutor.cpp

```diff
@@ -41,132 +49,197 @@ using Stats = NoStats;
+                    vectorComponentsCount));
   }
 }
 
-void EnumerateNearVectorsExecutor::searchResults(
-    std::vector<float>& inputRowsJoined) {
+void EnumerateNearVectorsExecutor::searchResults() {
   auto* vectorIndex = dynamic_cast<RocksDBVectorIndex*>(_infos.index.get());
   TRI_ASSERT(vectorIndex != nullptr);
 
-  std::vector<float> convertedInputRows(_inputRows.size());
   RocksDBMethods* mthds =
       RocksDBTransactionState::toMethods(&_trx, _collection->id());
 
   std::tie(_labels, _distances) = vectorIndex->readBatch(
-      inputRowsJoined, mthds, &_trx, _collection->getCollection(),
-      _inputRows.size(), _infos.topK);
-
-  _initialized = true;
-  _currentProcessedResultCount = 0;
+      _inputRowConverted, mthds, &_trx, _collection->getCollection(), 1,
+      _infos.getNumberOfResults());
+  LOG_INTERNAL << "Results: " << _labels << " and distances: " << _distances;
 }
 
 void EnumerateNearVectorsExecutor::fillOutput(OutputAqlItemRow& output) {
   auto const docOutId = _infos.outDocumentIdReg;
   auto const distOutId = _infos.outDistancesReg;
 
-  auto inputRowIterator = _inputRows.begin();
   while (!output.isFull() &&
-         _currentProcessedResultCount < _inputRows.size() * _infos.topK) {
+         _currentProcessedResultCount < _infos.getNumberOfResults()) {
     // there are no results anymore for this input, so we can skip to next input
     // row
     if (_labels[_currentProcessedResultCount] == -1) {
-      _currentProcessedResultCount =
-          (_currentProcessedResultCount / _infos.topK + 1) * _infos.topK;
-      ++inputRowIterator;
-      continue;
+      _currentProcessedResultCount = _infos.getNumberOfResults();
+      break;
     }
     output.moveValueInto(
-        docOutId, *inputRowIterator,
+        docOutId, _inputRow,
         AqlValueHintUInt(_labels[_currentProcessedResultCount]));
     output.moveValueInto(
-        distOutId, *inputRowIterator,
+        distOutId, _inputRow,
         AqlValueHintDouble(_distances[_currentProcessedResultCount]));
     output.advanceRow();
 
     ++_currentProcessedResultCount;
-    if (_currentProcessedResultCount % _infos.topK == 0) {
-      ++inputRowIterator;
-    }
+  }
+
+  if (_currentProcessedResultCount == _infos.getNumberOfResults()) {
+    _resultsAreProcessed = true;
   }
 }
 
 std::tuple<ExecutorState, NoStats, AqlCall>
 EnumerateNearVectorsExecutor::produceRows(AqlItemBlockInputRange& inputRange,
                                           OutputAqlItemRow& output) {
-  NoStats stats;
+  LOG_INTERNAL << "hasDataRow: " << inputRange.hasDataRow();
+  while (!output.isFull()) {
+    if (!_initialized) {
+      if (!_inputRow.isInitialized()) {
+        fillInput(inputRange);
+      }
+
+      searchResults();
+    }
+    fillOutput(output);
+    if (_currentProcessedResultCount == _infos.getNumberOfResults()) {
+      inputRange.advanceDataRow();
+      _initialized = false;
+      break;
+    }
```

Comment:
No, it advances to a new row after we are done with the current. Maybe just a general flow of the executor, or how I think it works:

`producesRows`:
1. Check if we can write to output
2. Check if we have already processed one input, if not call `fillInput` and parse the input (query point)
3. Now run faiss readBatch and get results for first input
4. For every result for the current input write it to output unless we encounter -1, meaning the results are invalid so we set `_currentProcessedResultCount` to `_infos.getNumberOfResults()` to avoid outputting anything. Once those two are equal we can go to next input

Reply:
Yes, but `fillInput` calls `inputRange.nextDataRow()`, which already advances the iterator on the input range. There's no need to call `advanceDataRow` again after you're finished with that row; just calling `nextDataRow()` again will give the next row.